# __Introduction to finpy-tse Python Module__
<font color='green'>__A Free Python Module to Access Tehran Stock Exchange Historical and Real-Time Data__</font>
## <font color='green'>__ALI RAHIMI__</font>

#### __Motivation:__ 

Data is new oil but extracting the data, is not as easy as it seems, like extracting the oil! When I started learning algorithmic trading and quantitative portfolio management approaches and models, my main barrier in going forward and examining these ideas was hardship in accessing the data. In the absence of easily accessible APIs, there were a lot of alternative ways to access the data manually or using Python, some of them was even premium but still imperfect, not enough for my requirements and not satisfactory. I had the tools for exploring the data but not the data itself! This was my main motivation for developing this module, maybe I would be able to help in removing this barrier from new learners's path, so that, they could focus on the data analysis and machine learning tools and do not worry about the data. __So, helping other people who started this way recently, is may gain from making this module, publicly available for free.__

#### __Scope:__  
    
The scope of this module is just __getting the data into your python enviroment__. Using this module, you will have most of the data, you may need for your research or development purposes. The finpy-tse module will provide you the __historical and real-time data of the Tehran Stock Exchange__ and I am pretty sure that you can leverage module-provided data to build any other specific form of price-oriented data. In this notebook, we will explain the module functions and you will learn how to use them to access the data. 
    
#### __Co-Developers:__

Some people helped me in developing this module and I would like to thank them, especially __Mr. Rasool Gafoori__ who _helped a lot in making module more faster and flexible_. I also would like to thank __Ms. M. Nourahmadi__ for her _comments from the user perspective and contribution for preparing module Farsi handbook_.


#### __Contact Info.:__

If you have comments or other requirements that is not covered in this version of finpy-tse, you can send me email or join our __Telegram Channel (@FinPy)__ and send your comments to admins. We will try to cover your requirements in the next verions of the module. I also invite you to join our Telegram Channel to know about module updates and Farsi tutorials.

Email: a.rahimi.aut@gmail.com

In [1]:
import finpy_tse as fpy

---------------------------------------------
## __1. Reading Single Stock Data__

Most people do their analysis based on stock daily data. In this section, we will introduce finpy-tse module functions, you can use, to access daily history of a single stock.

### __1.1. Get_Price_History() Function__
<font color='green'>__Access to Stock Daily Historical Price Data (= سابقه قیمت سهم)__</font>

This function reads a given stock price history in a given period of time starting from start_date till end_date (both are included, if they are trading days). The output is a Pandas' dataframe with some columns. The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __stock__ | String (in Farsi) | Ticker or name of firm in farsi | نماد یا نام کامل شرکت (سهم) مورد نظر به فارسی|
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه قیمت برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه قیمت برگردانده شده توسط تابع |
| __ignore_date__ | True/Flase | if True, the fucntion does not consider start and end date and gets all available history | توجه یا عدم توجه به تاریخ شروع و پایان در سابقه قیمت برگردانده شده توسط تابع |
| __adjust_price__ | True/Flase | if True, the fucntion returns adjusted Open, High, Low, Close and Final prices in new columns  |تعدیل یا عدم تعدیل قیمت برای افزایش سرمایه و پرداخت سود نقدی |
| __show_weekday__ | True/Flase | if True, the fucntion shows weekdays in a separate column |نمایش یا عدم نمایش نام روزهای هفته در خروجی تابع در یک ستون مجزا |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |نمایش یا عدم نمایش تاریخ میلادی در یک ستون مجزا علاوه بر تاریخ شمسی |

Returned dataframe's columns are:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی |
| __Date__ | International date | تاریخ میلادی |
| __Weekday__ | Name of weekdays | نام روز هفته |
| __Open__ | Day Open Price | قیمت باز شدن روز |
| __High__ | Day High Price | بالاترین قیمت روز |
| __Low__ | Day Low Price | پایین ترین قیمت روز |
| __Close__ | Day Close Price | آخرین قیمت روز |
| __Final__ | Day Final Price | قیمت پایانی روز |
| __Volume__ | Day Traded Volume| حجم معامله شده در آن روز |
| __Value__ | Day Traded Value (Rial) |ارزش معاملات سهم در آن روز (ریال) |
| __No__ | Number of Trades in that Day | تعداد معاملات انجام شده در آن روز |
| __Ticker__ | Ticker of the firm | نماد فارسی شرکت |
| __Name__ | Farsi Full Name of the firm | نام کامل فارسی شرکت |
| __Part__ | Data Part Number | شماره پارت دیتا، دیتای برخی سهم ها به دلیل تغییر بازار در چند پارت قابل دسترسی است |
| __Adj Open__ | Adjusted Day Open Price |  Open قیمت تعدیل شده  |
| __Adj High__ | Adjusted Day High Price | High قیمت تعدیل شده |
| __Adj Low__ | Adjusted Day Low Price | Low قیمت تعدیل شده |
| __Adj Close__ | Adjusted Day Close Price | Close قیمت تعدیل شده |
| __Adj Final__ | Adjusted Day Final Price | Final قیمت تعدیل شده |

In [2]:
df_price_history = fpy.Get_Price_History(stock = 'ذوب آهن اصفهان', start_date = '1399-12-15', end_date='1401-01-01',\
                                         ignore_date = True, adjust_price = True, show_weekday = True, double_date = True)
df_price_history

Date    Weekday  Open  High   Low  Close  Final     Volume  \
J-Date                                                                        
1390-11-19 2012-02-08  Wednesday  1304  1304  1304   1304   1304  248387500   
1390-11-23 2012-02-12     Sunday  1369  1369  1320   1369   1365   75848282   
1390-11-24 2012-02-13     Monday  1379  1397  1301   1311   1350   23435322   
1390-11-25 2012-02-14    Tuesday  1320  1324  1293   1317   1304   12352753   
1390-11-26 2012-02-15  Wednesday  1311  1324  1282   1290   1301    9402675   
...               ...        ...   ...   ...   ...    ...    ...        ...   
1400-10-29 2022-01-19  Wednesday  2725  2820  2725   2758   2740  228466128   
1400-11-02 2022-01-22   Saturday  2795  2850  2603   2603   2651  107941407   
1400-11-03 2022-01-23     Sunday  2526  2588  2519   2519   2522   31395792   
1400-11-06 2022-01-26  Wednesday  2550  2570  2396   2545   2480  115628493   
1400-11-09 2022-01-29   Saturday  2551  2604  2500   2604   2583   58679462   

                   Value    No Ticker                   Name  Part  Adj Open  \
J-Date                                                                         
1390-11-19  323897300000  2061    ذوب  سهامی ذوب آهن  اصفهان     2       164   
1390-11-23  103516867261  4290    ذوب  سهامی ذوب آهن  اصفهان     2       172   
1390-11-24   31646815791  2189    ذوب  سهامی ذوب آهن  اصفهان     2       173   
1390-11-25   16101932134  1051    ذوب  سهامی ذوب آهن  اصفهان     2       166   
1390-11-26   12234238887   680    ذوب  سهامی ذوب آهن  اصفهان     2       165   
...                  ...   ...    ...                    ...   ...       ...   
1400-10-29  626097659001  7884    ذوب         ذوب آهن اصفهان     1      2725   
1400-11-02  286200272904  4367    ذوب         ذوب آهن اصفهان     1      2795   
1400-11-03   79166777666   990    ذوب         ذوب آهن اصفهان     1      2526   
1400-11-06  286812147802  4675    ذوب         ذوب آهن اصفهان     1      2550   
1400-11-09  151546331963  2331    ذوب         ذوب آهن اصفهان     1      2551   

            Adj High  Adj Low  Adj Close  Adj Final  
J-Date                                               
1390-11-19       164      164        164        164  
1390-11-23       172      166        172        171  
1390-11-24       176      163        165        170  
1390-11-25       166      162        165        164  
1390-11-26       166      161        162        163  
...              ...      ...        ...        ...  
1400-10-29      2820     2725       2758       2740  
1400-11-02      2850     2603       2603       2651  
1400-11-03      2588     2519       2519       2522  
1400-11-06      2570     2396       2545       2480  
1400-11-09      2604     2500       2604       2583  

[2140 rows x 18 columns]

In [3]:
df_price_history['1399':]

Date    Weekday  Open  High   Low  Close  Final     Volume  \
J-Date                                                                        
1399-01-05 2020-03-24    Tuesday  2700  2749  2553   2599   2593  210506642   
1399-01-06 2020-03-25  Wednesday  2611  2703  2600   2685   2666  213584985   
1399-01-09 2020-03-28   Saturday  2727  2799  2673   2799   2755  274807499   
1399-01-10 2020-03-29     Sunday  2850  2892  2760   2810   2849  317918782   
1399-01-11 2020-03-30     Monday  2831  2935  2761   2895   2858  273617729   
...               ...        ...   ...   ...   ...    ...    ...        ...   
1400-10-29 2022-01-19  Wednesday  2725  2820  2725   2758   2740  228466128   
1400-11-02 2022-01-22   Saturday  2795  2850  2603   2603   2651  107941407   
1400-11-03 2022-01-23     Sunday  2526  2588  2519   2519   2522   31395792   
1400-11-06 2022-01-26  Wednesday  2550  2570  2396   2545   2480  115628493   
1400-11-09 2022-01-29   Saturday  2551  2604  2500   2604   2583   58679462   

                   Value     No Ticker                   Name  Part  Adj Open  \
J-Date                                                                          
1399-01-05  545891694600  19947    ذوب  سهامی ذوب آهن  اصفهان     2      2659   
1399-01-06  569426083670  19864    ذوب  سهامی ذوب آهن  اصفهان     2      2571   
1399-01-09  757155502726  22043    ذوب  سهامی ذوب آهن  اصفهان     2      2685   
1399-01-10  905895747263  29495    ذوب  سهامی ذوب آهن  اصفهان     2      2807   
1399-01-11  781877324354  21910    ذوب  سهامی ذوب آهن  اصفهان     2      2788   
...                  ...    ...    ...                    ...   ...       ...   
1400-10-29  626097659001   7884    ذوب         ذوب آهن اصفهان     1      2725   
1400-11-02  286200272904   4367    ذوب         ذوب آهن اصفهان     1      2795   
1400-11-03   79166777666    990    ذوب         ذوب آهن اصفهان     1      2526   
1400-11-06  286812147802   4675    ذوب         ذوب آهن اصفهان     1      2550   
1400-11-09  151546331963   2331    ذوب         ذوب آهن اصفهان     1      2551   

            Adj High  Adj Low  Adj Close  Adj Final  
J-Date                                               
1399-01-05      2707     2514       2559       2553  
1399-01-06      2662     2560       2644       2625  
1399-01-09      2756     2632       2756       2713  
1399-01-10      2848     2718       2767       2806  
1399-01-11      2890     2719       2851       2814  
...              ...      ...        ...        ...  
1400-10-29      2820     2725       2758       2740  
1400-11-02      2850     2603       2603       2651  
1400-11-03      2588     2519       2519       2522  
1400-11-06      2570     2396       2545       2480  
1400-11-09      2604     2500       2604       2583  

[424 rows x 18 columns]

### __1.2. Get_RI_History() Function__
<font color='green'>__Access to Stock Historical Retail-Institutional Traders Data (= سابقه حقیقی-حقوقی سهم)__</font>

In this function, __R__ stands for retail traders and __I__ stands for institutional traders. This function returns retail and institutional traders trades history in daily time frame, for a given stock, in a given period between start and end date. The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __stock__ | String (in Farsi) | Ticker or name of firm in farsi | نماد یا نام کامل شرکت (سهم) مورد نظر به فارسی|
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه حقیقی-حقوقی برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه حقیقی-حقوقی برگردانده شده توسط تابع |
| __ignore_date__ | True/Flase | if True, the fucntion does not consider start and end date and gets all available history | توجه یا عدم توجه به تاریخ شروع و پایان در سابقه حقیقی-حقوقی برگردانده شده توسط تابع |
| __show_weekday__ | True/Flase | if True, the fucntion shows weekdays in a separate column |نمایش یا عدم نمایش نام روزهای هفته در خروجی تابع در یک ستون مجزا |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |نمایش یا عدم نمایش تاریخ میلادی در یک ستون مجزا علاوه بر تاریخ شمسی |

Returned dataframe's columns are:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی |
| __Date__ | International date | تاریخ میلادی |
| __Weekday__ | Name of weekdays | نام روز هفته |
| __No_Buy_R__ | Number of buy trades by retail traders | تعداد خرید حقیقی |
| __No_Buy_I__ | Number of buy trades by institutional traders | تعداد خرید حقوقی |
| __No_Sell_R__ | Number of sell trades by retail traders | تعداد فروش حقیقی |
| __No_Sell_I__ | Number of sell trades by institutional traders | تعداد فروش حقوقی |
| __Vol_Buy_R__ | Volume of buy trades by retail traders | حجم خرید حقیقی |
| __Vol_Buy_I__ | Volume of buy trades by institutional traders | حجم خرید حقوقی |
| __Vol_Sell_R__ | Volume of sell trades by retail traders | حجم فروش حقیقی |
| __Vol_Sell_I__ | Volume of sell trades by institutional traders | حجم فروش حقوقی |
| __Val_Buy_R__ | Value of buy trades by retail traders | ارزش خرید حقیقی |
| __Val_Buy_I__ | Value of buy trades by institutional traders | ارزش خرید حقوقی |
| __Val_Sell_R__ | Value of sell trades by retail traders | ارزش فروش حقیقی |
| __Val_Sell_I__ | Value of sell trades by institutional traders | ارزش فروش حقوقی |
| __Ticker__ | Ticker of the firm | نماد فارسی شرکت |
| __Name__ | Farsi Full Name of the firm | نام کامل فارسی شرکت |
| __Part__ | Data Part Number | شماره پارت دیتا، دیتای برخی سهم ها به دلیل تغییر بازار در چند پارت قابل دسترسی است |

In [4]:
df_ri_history = fpy.Get_RI_History(stock = 'پترول', start_date = '1396-01-01', end_date='1400-12-29',\
                                   ignore_date = False, show_weekday = False, double_date = False)
df_ri_history

No_Buy_R  No_Buy_I  No_Sell_R  No_Sell_I  Vol_Buy_R  Vol_Buy_I  \
J-Date                                                                       
1396-01-05        73         0         34          2    2268915          0   
1396-01-06        68         0         54          1    4259110          0   
1396-01-07        49         0        104          1    3160371          0   
1396-01-08        37         0        136          0    5757038          0   
1396-01-09       142         2        207          0    6088665     300000   
...              ...       ...        ...        ...        ...        ...   
1400-11-03      1548         6       1028          1   67469072   18469737   
1400-11-04      1151         6        907          3   46743910    4953625   
1400-11-05       907         5       1336          3   68151202   11300000   
1400-11-06       441         0        893          5   48575736          0   
1400-11-09       266         1        455          2   34247778   10000000   

            Vol_Sell_R  Vol_Sell_I     Val_Buy_R    Val_Buy_I    Val_Sell_R  \
J-Date                                                                        
1396-01-05     1679576      589339    2305515221            0    1705804930   
1396-01-06     1357863     2901247    4355141482            0    1391938729   
1396-01-07     3060371      100000    3289575734            0    3184675734   
1396-01-08     5757038           0    6188076443            0    6188076443   
1396-01-09     6388665           0    6639660679    324850000    6964510679   
...                ...         ...           ...          ...           ...   
1400-11-03    85924819       13990  131091858132  35813650873  166878088605   
1400-11-04    50306502     1391033   87325457569   9155005375   93896307684   
1400-11-05    76885835     2565367  123244363390  20771462906  139377520587   
1400-11-06    45951910     2623826   92433760394            0   87440625516   
1400-11-09    18689152    25558626   68427060444  19980000000   37340925696   

             Val_Sell_I Ticker                         Name Part  
J-Date                                                            
1396-01-05    599710291  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1396-01-06   2963202753  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1396-01-07    104900000  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1396-01-08            0  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1396-01-09            0  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
...                 ...    ...                          ...  ...  
1400-11-03     27420400  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-11-04   2584155260  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-11-05   4638305709  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-11-06   4993134878  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-11-09  51066134748  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  

[1085 rows x 15 columns]

---------------------------------------------
## __2. Reading Single Stock Intraday Data__

Some people develop their trading strategies based on intra-day data. In this section, we will introduce finpy-tse module functions, you can use to access this type of data.

### __2.1. Get_IntradayTrades_History() Function__
<font color='green'>__Access to Stock Historical Intra-Day Trades Data (= سابقه ریز معاملات سهم)__</font>

Tehran Stock Exchange is open from 8:45 to 12:30. Orders are executed from 9:00 to 12:30 and from 8:45 to 9:00 is pre-opening period. In pre-opening period, you can place orders but they are not executed until 9:00. This function returns every signle executed trade with its execution time, volume and price. You can use this type of output to create price bars in different time frames, tick charts, volume bars, information bars, etc. as mentioned is Marcos Lopez Deprada's book, "Advances in Financial Machine Learning". 

The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __stock__ | String (in Farsi) | Ticker or name of firm in farsi | نماد یا نام کامل شرکت (سهم) مورد نظر به فارسی|
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه ریز معاملات برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه ریز معاملات برگردانده شده توسط تابع |
| __jalali_date__ | True/False | If True, the function returns Jalali Date, otherwise it returns international date |نمایش تاریخ به صورت شمسی یا میلادی در خروجی تابع |
| __combined_datetime__ | True/False | If True, the function combines date and time and shows as single column |تلفیق یا عدم تلفیق زمان با تاریخ |
| __show_progress__ | True/False | If True, shows the % progress in gathering and processing data |نمایش یا عدم نمایش درصد پیشرفت اطلاعات جمع آوری و پردازش شده |

Returned dataframe's columns are:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی انجام معامله |
| __Date__ | International date | تاریخ میلادی انجام معامله |
| __Time__ | Day Time | ساعت انجام معامله |
| __J-DateTime__ | Jalali date and time |  تاریخ شمسی و ساعت انجام معامله |
| __DateTime__ | International date and time | تاریخ میلادی و ساعت انجام معامله |
| __Volume__ | Traded Volume | حجم معامله انجام شده در آن تاریخ و ساعت |
| __Price__ | Trade Price | قیمت معامله انجام شده در آن تاریخ و ساعت |

__Important Note:__ Combined Jalali DateTime takes more time to run. So, if you are interested in that sort of output, be patient when running the function.

In [5]:
df_intraday = fpy.Get_IntradayTrades_History(stock = 'فولاد', start_date = '1400-09-20', end_date='1400-10-20',
                                             jalali_date = True, combined_datatime = False, show_progress = True)
df_intraday

Progress :  100.0 %


Volume  Price
J-Date     Time                   
1400-09-20 09:00:16   37000  10540
           09:00:16    7430  10540
           09:00:16    6308  10540
           09:00:16    6000  10540
           09:00:16    5213  10540
...                     ...    ...
1400-10-20 12:29:56     510  10080
           12:29:57    1600  10060
           12:29:59    1976  10080
           12:29:59    5500  10060
           12:29:59     500  10060

[169683 rows x 2 columns]

### __2.2. Get_IntradayOB_History() Function__
<font color='green'>__Intra-day, Order Book (OB), Historical Data for a Single Stock (= سابقه دفتر معاملات سهم)__</font>

Tehran Stock Exchange has a queue-based order book (OB). Five first rows of this OB is accessible for public and this function gets that data for previous days. Remember, this is not real time data, but it is for the past days. You can use this data to create your own dataset and develop trading strategies based on what happens in supply and demand.

The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __stock__ | String (in Farsi) | Ticker or name of firm in farsi | نماد یا نام کامل شرکت (سهم) مورد نظر به فارسی|
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه دفتر معاملات برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه دفتر معاملات برگردانده شده توسط تابع |
| __jalali_date__ | True/False | If True, the function returns Jalali Date, otherwise it returns international date |نمایش تاریخ به صورت شمسی یا میلادی در خروجی تابع |
| __combined_datetime__ | True/False | If True, the function combines date and time and shows as single column |تلفیق یا عدم تلفیق زمان با تاریخ |
| __show_progress__ | True/False | If True, shows the % progress in gathering and processing data |نمایش یا عدم نمایش درصد پیشرفت اطلاعات جمع آوری و پردازش شده |

Returned dataframe's columns are:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی انجام معامله |
| __Date__ | International date | تاریخ میلادی انجام معامله |
| __Time__ | Day Time | ساعت انجام معامله |
| __J-DateTime__ | Jalali date and time |  تاریخ شمسی و ساعت انجام معامله |
| __DateTime__ | International date and time | تاریخ میلادی و ساعت انجام معامله |
| __Depth__ |Order Book Depth (1-5)| عمق دفتر سفارشات (بهترین مظنه در عمق 1) |
| __Sell-No__ |Number of People Willing to Sell at Given Price| تعداد افراد حاضر به فروش در آن مظنه فروش |
| __Sell-Vol__ |Day Open Price| حجم تجمعی فروشندگان در آن مظنه فروش |
| __Sell-Price__ |Offered Sell Price| قیمت پیشنهادی فروشندگان در آن مظنه فروش |
| __Buy-Price__ |Offered Buy Price| قیمت پیشنهادی خریداران در آن مظنه خرید |
| __Buy-Vol__ |Ready Buy Volume at Given Price| حجم تجمعی خریداران در آن مظنه خرید  |
| __Buy-No__ |Number of People Willing to Buy at Given Price| تعداد افراد حاضر به خرید در آن مظنه خرید |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز برای معامله در آن روز  |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز برای معامله در آن روز  |

__Important Note:__ Combined Jalali DateTime takes more time to run. So, if you are interested in that sort of output, be patient when running the function.

In [6]:
df_ob = fpy.Get_IntradayOB_History(stock = 'ذوب', start_date = '1400-08-15', end_date='1400-09-15',
                                   jalali_date = False, combined_datatime = True, show_progress = True)
df_ob

Progress :  100.0 %


Sell_No Sell_Vol  Sell_Price  Buy_Price Buy_Vol  \
DateTime            Depth                                                   
2021-11-06 08:45:02 1           7    15420      4091.0     4091.0   50000   
                    2           1    60000      4118.0     3964.0   25135   
                    3           1    50000      4120.0     3813.0    1500   
                    4           1     5000      4180.0     3800.0    1400   
                    5           3    21364      4189.0     3750.0   30000   
...                           ...      ...         ...        ...     ...   
2021-12-06 12:29:54 2           1     1000      3443.0     3436.0  101759   
                    3           2     7670      3444.0     3435.0  472570   
                    4           3    23991      3450.0     3434.0  112710   
                    5           3     8500      3459.0     3433.0  116500   
2021-12-06 12:29:58 1           1     1069      3440.0     3437.0   27066   

                          Buy_No  Day_LL  Day_UL  
DateTime            Depth                         
2021-11-06 08:45:02 1          1  3703.0  4091.0  
                    2          1  3703.0  4091.0  
                    3          1  3703.0  4091.0  
                    4          1  3703.0  4091.0  
                    5          1  3703.0  4091.0  
...                          ...     ...     ...  
2021-12-06 12:29:54 2          2  3293.0  3639.0  
                    3         12  3293.0  3639.0  
                    4          3  3293.0  3639.0  
                    5          3  3293.0  3639.0  
2021-12-06 12:29:58 1          3  3293.0  3639.0  

[179101 rows x 8 columns]

### __2.3. Get_Queue_History() Function__
<font color='green'>__Market Closing Queue History for a Given Stock and Period (= سابقه ارزش صف خرید یا فروش سهم در زمان بسته شدن بازار)__</font>

Tehran Stock Exchange has a queue-based order book (OB). Most of the time, when the market is closed, there are some unfilled orders in the order book and behind the buy/sell queues, which shows some sort of supply/demand accumulation for that stock. You can include these data in your analysis for developing better and more realistic trading strategies, as TSE is not liquid enough to get in/out of a trade, whenever your trading strategy wants. This function will provide you the historical data of buy/sell queues for a given stock, in a given period, when the market is closed around 12:30.

The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __stock__ | String (in Farsi) | Ticker or name of firm in farsi | نماد یا نام کامل شرکت (سهم) مورد نظر به فارسی|
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه ارزش صف های خرید یا فروش برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه ارزش صف های خرید یا فروش برگردانده شده توسط تابع |
| __show_per_capita__ | True/Flase | if True, the fucntion returns per capita value of queue |نمایش یا عدم نمایش سرانه سفارشات در صف خرید یا فروش |
| __show_weekday__ | True/Flase | if True, the fucntion shows weekdays in a separate column |نمایش یا عدم نمایش نام روزهای هفته در خروجی تابع در یک ستون مجزا |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |نمایش یا عدم نمایش تاریخ میلادی در یک ستون مجزا علاوه بر تاریخ شمسی |
| __show_progress__ | True/False | If True, shows the % progress in gathering and processing data |نمایش یا عدم نمایش درصد پیشرفت اطلاعات جمع آوری و پردازش شده |

Returned dataframe's columns are:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی |
| __Date__ | International date | تاریخ میلادی |
| __Weekday__ | Name of weekdays | نام روز هفته |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز روز جاری |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز روز جاری |
| __Value__ |Value of Executed Trades| ارزش معاملات درآن روز |
| __Time__ |Time of the latest available order book data before market close| زمان اطلاعات آخرین دفتر معاملات در دسترس قبل از بسته شدن بازار |
| __BQ-Value__ |Buy Queue Value (Rial) | ارزش صف خرید در انتهای آن روز و در زمان بسته شدن بازار (ریال) |
| __SQ-Value__ |Sell Queue Value (Rial)| ارزش صف فروش در انتهای آن روز و در زمان بسته شدن بازار (ریال) |
| __BQPC__ |Buy Queue Per-Capita (Rial)| ارزش سرانه صف خرید در انتهای آن روز و در زمان بسته شدن بازار (ریال) |
| __SQPC__ |Sell Queue Per-Capita (Rial)| ارزش سرانه صف فروش در انتهای آن روز و در زمان بسته شدن بازار (ریال) |


In [7]:
df_bq_sq_val = fpy.Get_Queue_History(stock = 'خساپا', start_date = '1398-04-15', end_date='1398-05-05', show_per_capita = True,\
                                     show_weekday = True, double_date = True, show_progress = True)
df_bq_sq_val

Progress :  100.0 %


Date    Weekday Day_UL Day_LL         Value      Time  \
J-Date                                                                   
1398-04-15 2019-07-06   Saturday   1712   1892  154180079727  12:29:55   
1398-04-16 2019-07-07     Sunday   1636   1808  326308143187  12:30:00   
1398-04-17 2019-07-08     Monday   1585   1751  334241736666  12:30:00   
1398-04-18 2019-07-09    Tuesday   1581   1747  213839121262  12:29:59   
1398-04-19 2019-07-10  Wednesday   1533   1693  285226432296  12:29:57   
1398-04-22 2019-07-13   Saturday   1579   1745  191800675506  12:30:00   
1398-04-23 2019-07-14     Sunday   1658   1832  131450536967  12:29:58   
1398-04-24 2019-07-15     Monday   1598   1766  199847905825  12:29:58   
1398-04-25 2019-07-16    Tuesday   1527   1687   95963836719  12:29:49   
1398-04-26 2019-07-17  Wednesday   1459   1611  355441723801  12:29:59   
1398-04-29 2019-07-20   Saturday   1450   1602  570068932799  12:30:00   
1398-05-02 2019-07-24  Wednesday   1468   1622   63696557989  12:30:00   
1398-05-05 2019-07-27   Saturday   1458   1610  112900264787  12:30:00   

               BQ_Value      SQ_Value      BQPC       SQPC  
J-Date                                                      
1398-04-15            0   79851073184         0  114072962  
1398-04-16            0             0         0          0  
1398-04-17            0             0         0          0  
1398-04-18            0             0         0          0  
1398-04-19  18674067652             0  94313473          0  
1398-04-22            0             0         0          0  
1398-04-23            0     641047462         0  128209492  
1398-04-24            0  133539342106         0  130536991  
1398-04-25            0  177596055300         0  113989766  
1398-04-26            0             0         0          0  
1398-04-29            0             0         0          0  
1398-05-02            0             0         0          0  
1398-05-05            0             0         0          0

---------------------------------------------
## __3. Reading Market Indices Historical Data__
<font color='green'>__شاخص های بازار بورس تهران__</font>

Market indices play an important role in performace evaluation and developing some investment or trading strategies. For example, most people like to calculate a __portfolio or stock's Beta coefficient__ with respect to overall market index. In this section we will introduce finpy-tse module's functions, you could use to access to Tehran Stock Exchange indices.

### __3.1. Get_CWI_History() Function__
<font color='green'>__Access to Cap-Weighted Index (CWI) Historical Data (= شاخص کل بورس)__</font>

You can use the cap-weighted index (CWI) to check if a portfolio or stock is ahead of the market or behind it, in term of performance. It can also be used in CAPM and other factor models like Fama-French and Carhart as market factor of the model. The function inputs are:

| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | :- | -: |
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) | تاریخ شمسی شروع سابقه شاخص برگردانده شده توسط تابع |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |تاریخ شمسی پایان سابقه شاخص برگردانده شده توسط تابع |
| __ignore_date__ | True/Flase | if True, the fucntion does not consider start and end date and gets all available history | توجه یا عدم توجه به تاریخ شروع و پایان در سابقه شاخص برگردانده شده توسط تابع |
| __just_adj_close__ | True/Flase | if True, the fucntion just returns the adjusted close column |تعیین اینکه آیا فقط آخرین قیمت تعدیل شده شاخص برگردانده شود یا همه اطلاعات |
| __show_weekday__ | True/Flase | if True, the fucntion shows weekdays in a separate column |نمایش یا عدم نمایش نام روزهای هفته در خروجی تابع در یک ستون مجزا |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |نمایش یا عدم نمایش تاریخ میلادی در یک ستون مجزا علاوه بر تاریخ شمسی |

Returned dataframe's columns are:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی |
| __Date__ | International date | تاریخ میلادی |
| __Weekday__ | Name of weekdays | نام روز هفته |
| __Open__ | Index Day Open Value | قیمت باز شدن شاخص در آن روز |
| __High__ | Index Day High Value | بالاترین قیمت شاخص در آن روز |
| __Low__ | Index Day Low Value | آخرین قیمت شاخص در آن روز |
| __Close__ | Index Day Close Value | آخرین قیمت روز |
| __Adj Close__ | Index Day Adjusted Close Value | آخرین قیمت تعدیل شده شاخص در آن روز |
| __Volume__ | Traded Volume of all Stocks in the Index| حجم معاملات همه سهم هایی که در شاخص هستند |

__Important Note:__ You can also use the following setting to get only the adj-close history of the index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [8]:
df_cwi_history = fpy.Get_CWI_History(start_date = '1395-01-01', end_date='1400-12-29', ignore_date = False, 
                                     just_adj_close = True, show_weekday = True, double_date = True)
df_cwi_history

Date    Weekday  Adj Close
J-Date                                     
1395-01-07 2016-03-26   Saturday    81200.3
1395-01-08 2016-03-27     Sunday    81261.0
1395-01-09 2016-03-28     Monday    80935.7
1395-01-10 2016-03-29    Tuesday    80561.3
1395-01-11 2016-03-30  Wednesday    81480.4
...               ...        ...        ...
1400-11-03 2022-01-23     Sunday  1256512.8
1400-11-04 2022-01-24     Monday  1234707.7
1400-11-05 2022-01-25    Tuesday  1207926.7
1400-11-06 2022-01-26  Wednesday  1229368.4
1400-11-09 2022-01-29   Saturday  1258995.1

[1415 rows x 3 columns]

### __3.2. Get_EWI_History() Function__
<font color='green'>__Access to Equal-Weighted Index (EWI) Historical Data (= شاخص کل هم وزن بورس)__</font>

You can use the equal-weighted index (EWI) to check if a portfolio or stock is ahead of the market or behind it, in term of performance. It can also be used to understand if small-cap stocks performance is better than large-cap stocks, or not. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [9]:
df_ewi_history = fpy.Get_EWI_History(start_date = '1395-01-01', end_date='1400-12-29', ignore_date = False, 
                                     just_adj_close = False, show_weekday = True, double_date = True)
df_ewi_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1395-01-07 2016-03-26   Saturday   13347.0   13450.0   13334.0   13450.0   
1395-01-08 2016-03-27     Sunday   13468.0   13484.0   13461.0   13462.0   
1395-01-09 2016-03-28     Monday   13458.0   13474.0   13455.0   13473.0   
1395-01-10 2016-03-29    Tuesday   13481.0   13484.0   13467.0   13470.0   
1395-01-11 2016-03-30  Wednesday   13498.0   13600.0   13498.0   13599.0   
...               ...        ...       ...       ...       ...       ...   
1400-11-03 2022-01-23     Sunday  333959.0  333959.0  328203.0  328244.0   
1400-11-04 2022-01-24     Monday  327947.0  327947.0  323972.0  323973.0   
1400-11-05 2022-01-25    Tuesday  323632.0  323632.0  317066.0  317066.0   
1400-11-06 2022-01-26  Wednesday  317345.0  318759.0  317345.0  318758.0   
1400-11-09 2022-01-29   Saturday  319219.0  325138.0  319219.0  325134.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    13450.1   785545121  
1395-01-08    13462.3  1006304416  
1395-01-09    13473.0   677383208  
1395-01-10    13470.0  1491153697  
1395-01-11    13598.9   947224063  
...               ...         ...  
1400-11-03   328246.0  4622360998  
1400-11-04   323974.0  3361522069  
1400-11-05   317067.0  5442842578  
1400-11-06   318756.0  4489233096  
1400-11-09   325135.0  3360744518  

[1414 rows x 8 columns]

### __3.3. Get_CWPI_History() Function__
<font color='green'>__Access to Cap-Weighted Price Index (CWPI) Historical Data (= شاخص قیمت وزنی-ارزشی بورس)__</font>

Cap-Weighted Price Index (CWPI) is very similar to Cap-Weighted Index (CWI), except that in CWPI, dividents has no effect in index calculations and there is no adjustment for price changes, resulting from DPS. Please note that the historical data for CWPI index, in the form of OHLCV, is available from 1393-12-04, but adj-close historical data is available from 1387-09-14. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [10]:
cwpi_history = fpy.Get_CWPI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
cwpi_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1395-01-07 2016-03-26   Saturday   30479.0   30884.0   30479.0   30884.0   
1395-01-08 2016-03-27     Sunday   30966.0   30979.0   30905.0   30907.0   
1395-01-09 2016-03-28     Monday   30835.0   30835.0   30782.0   30782.0   
1395-01-10 2016-03-29    Tuesday   30747.0   30765.0   30640.0   30640.0   
1395-01-11 2016-03-30  Wednesday   30721.0   30997.0   30721.0   30990.0   
...               ...        ...       ...       ...       ...       ...   
1400-11-03 2022-01-23     Sunday  316124.0  316124.0  310107.0  310107.0   
1400-11-04 2022-01-24     Monday  309838.0  309838.0  304713.0  304713.0   
1400-11-05 2022-01-25    Tuesday  304511.0  304511.0  298105.0  298105.0   
1400-11-06 2022-01-26  Wednesday  299545.0  303394.0  299545.0  303394.0   
1400-11-09 2022-01-29   Saturday  304373.0  310687.0  304373.0  310684.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    30884.2   785545121  
1395-01-08    30906.2  1006304416  
1395-01-09    30782.5   677383208  
1395-01-10    30640.1  1491153697  
1395-01-11    30989.7   947224063  
...               ...         ...  
1400-11-03   310099.0  4622360998  
1400-11-04   304698.0  3361522069  
1400-11-05   298089.0  5442842578  
1400-11-06   303381.0  4489233096  
1400-11-09   310691.0  3360744518  

[1414 rows x 8 columns]

### __3.4. Get_EWPI_History() Function__
<font color='green'>__Access to Equal-Weighted Price Index (EWPI) Historical Data (= شاخص قیمت هم وزن بورس)__</font>

Equal-Weighted Price Index (EWPI) is very similar to Equal-Weighted Index (EWI), except that in EWPI, dividents has no effect in index calculations and there is no adjustment for price changes, resulting from DPS. Please note that the historical data for EWPI index, in the form of OHLCV, is available from 1393-12-05, but adj-close historical data is available from 1392-12-28. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [11]:
ewpi_history = fpy.Get_EWPI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = True, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
ewpi_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1393-12-05 2015-02-24    Tuesday    9198.0    9212.0    9182.0    9182.0   
1393-12-06 2015-02-25  Wednesday    9183.0    9183.0    9109.0    9109.0   
1393-12-09 2015-02-28   Saturday    9112.0    9117.0    9090.0    9090.0   
1393-12-10 2015-03-01     Sunday    9014.0    9030.0    9009.0    9030.0   
1393-12-11 2015-03-02     Monday    9034.0    9056.0    9034.0    9044.0   
...               ...        ...       ...       ...       ...       ...   
1400-11-03 2022-01-23     Sunday  209713.0  209713.0  206098.0  206124.0   
1400-11-04 2022-01-24     Monday  205937.0  205937.0  203442.0  203442.0   
1400-11-05 2022-01-25    Tuesday  203137.0  203137.0  199016.0  199016.0   
1400-11-06 2022-01-26  Wednesday  199191.0  200079.0  199191.0  200078.0   
1400-11-09 2022-01-29   Saturday  200367.0  204083.0  200367.0  204080.0   

            Adj Close      Volume  
J-Date                             
1393-12-05     9115.9   278521094  
1393-12-06     9033.8   271188153  
1393-12-09     9014.7   257849083  
1393-12-10     9030.4   475683577  
1393-12-11     9043.5   717632564  
...               ...         ...  
1400-11-03   206126.0  4622360998  
1400-11-04   203352.0  3361522069  
1400-11-05   199017.0  5442842578  
1400-11-06   200076.0  4489233096  
1400-11-09   204065.0  3360744518  

[1673 rows x 8 columns]

### __3.5. Get_FFI_History() Function__
<font color='green'>__Access to Free-Float Index (FFI) Historical Data (= TEFIX = شاخص آزاد شناور بورس)__</font>

Free-Float Index (FFI) is very similar to Cap-Weighted Index (CWI), except that in FFI, instead of using all released shares for a stock, we only use number of free float shares to calculate market cap. Please note that the historical data for FFI index, in the form of OHLCV, is available from 1388-04-09, but adj-close historical data is available from 1388-04-08. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [12]:
ffi_history = fpy.Get_FFI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                  just_adj_close = False, show_weekday = True, double_date = True)
ffi_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    92349.0    93869.0    92349.0    93869.0   
1395-01-08 2016-03-27     Sunday    94166.0    94204.0    93830.0    93830.0   
1395-01-09 2016-03-28     Monday    93424.0    93424.0    93262.0    93262.0   
1395-01-10 2016-03-29    Tuesday    93025.0    93052.0    92664.0    92676.0   
1395-01-11 2016-03-30  Wednesday    93026.0    94149.0    93026.0    94147.0   
...               ...        ...        ...        ...        ...        ...   
1400-11-03 2022-01-23     Sunday  1634660.0  1634660.0  1604330.0  1604350.0   
1400-11-04 2022-01-24     Monday  1602660.0  1602660.0  1580560.0  1580570.0   
1400-11-05 2022-01-25    Tuesday  1579720.0  1579720.0  1549330.0  1549330.0   
1400-11-06 2022-01-26  Wednesday  1559350.0  1579770.0  1559350.0  1579770.0   
1400-11-09 2022-01-29   Saturday  1586140.0  1619030.0  1586140.0  1619010.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    93869.0   785545121  
1395-01-08    93830.5  1006304416  
1395-01-09    93262.5   677383208  
1395-01-10    92676.3  1491153697  
1395-01-11    94147.3   947224063  
...               ...         ...  
1400-11-03  1604300.0  4622360998  
1400-11-04  1580570.0  3361522069  
1400-11-05  1549260.0  5442842578  
1400-11-06  1579680.0  4489233096  
1400-11-09  1619080.0  3360744518  

[1414 rows x 8 columns]

### __3.6. Get_MKT1I_History() Function__
<font color='green'>__Access to First Market Index (MKT1I) Historical Data (= شاخص بازار اول بورس)__</font>

In Tehran Stock Exchange's Bourse market, stocks has been divided into two sub-markets, first market and second market, based on their profitiblity, free float, liquidity and some other parameters. Each market has its own cap-weighted index. This function gets the first market index (MKT1I) data for you. Please note that the historical data for MKT1I index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-14. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [13]:
mkt1i_history = fpy.Get_MKT1I_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
mkt1i_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1395-01-07 2016-03-26   Saturday   56906.0   57896.0   56906.0   57897.0   
1395-01-08 2016-03-27     Sunday   58114.0   58145.0   57926.0   57932.0   
1395-01-09 2016-03-28     Monday   57754.0   57754.0   57636.0   57636.0   
1395-01-10 2016-03-29    Tuesday   57533.0   57533.0   57190.0   57192.0   
1395-01-11 2016-03-30  Wednesday   57302.0   57948.0   57302.0   57926.0   
...               ...        ...       ...       ...       ...       ...   
1400-11-03 2022-01-23     Sunday  973270.0  973270.0  953602.0  953602.0   
1400-11-04 2022-01-24     Monday  952674.0  952674.0  934932.0  934932.0   
1400-11-05 2022-01-25    Tuesday  934511.0  934511.0  912445.0  912445.0   
1400-11-06 2022-01-26  Wednesday  918310.0  931668.0  918310.0  931668.0   
1400-11-09 2022-01-29   Saturday  935739.0  958266.0  935739.0  958265.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    57896.6   627397539  
1395-01-08    57932.0   572942045  
1395-01-09    57635.6   440061076  
1395-01-10    57191.5  1027589626  
1395-01-11    57926.3   676272786  
...               ...         ...  
1400-11-03   953603.0  2289094777  
1400-11-04   934930.0  1898788274  
1400-11-05   912401.0  3388408324  
1400-11-06   931639.0  2481673500  
1400-11-09   958334.0  1750966776  

[1414 rows x 8 columns]

### __3.7. Get_MKT2I_History() Function__
<font color='green'>__Access to Second Market Index (MKT2I) Historical Data (= شاخص بازار دوم بورس)__</font>

In Tehran Stock Exchange's Bourse market, stocks has been divided into two sub-markets, first market and second market, based on their profitiblity, free float, liquidity and some other parameters. Each market has its own cap-weighted index. This function gets the second market index (MKT2I) data for you. Please note that the historical data for MKT2I index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-14. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [14]:
mkt2i_history = fpy.Get_MKT2I_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
mkt2i_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday   170302.0   171185.0   170302.0   171185.0   
1395-01-08 2016-03-27     Sunday   171268.0   171373.0   171268.0   171359.0   
1395-01-09 2016-03-28     Monday   171235.0   171235.0   171046.0   171057.0   
1395-01-10 2016-03-29    Tuesday   171081.0   171466.0   171081.0   171322.0   
1395-01-11 2016-03-30  Wednesday   172018.0   172787.0   172018.0   172787.0   
...               ...        ...        ...        ...        ...        ...   
1400-11-03 2022-01-23     Sunday  2470830.0  2470830.0  2427550.0  2427560.0   
1400-11-04 2022-01-24     Monday  2425760.0  2425760.0  2391760.0  2391760.0   
1400-11-05 2022-01-25    Tuesday  2389900.0  2389900.0  2347070.0  2347070.0   
1400-11-06 2022-01-26  Wednesday  2353910.0  2379340.0  2353910.0  2379330.0   
1400-11-09 2022-01-29   Saturday  2383710.0  2423530.0  2383710.0  2423470.0   

            Adj Close      Volume  
J-Date                             
1395-01-07   171185.0   158147582  
1395-01-08   171359.0   433362371  
1395-01-09   171057.0   237322132  
1395-01-10   171322.0   463564071  
1395-01-11   172787.0   270951277  
...               ...         ...  
1400-11-03  2427410.0  2333266221  
1400-11-04  2391850.0  1462733795  
1400-11-05  2346930.0  2054434254  
1400-11-06  2379190.0  2007559596  
1400-11-09  2423400.0  1609777742  

[1414 rows x 8 columns]

### __3.8. Get_INDI_History() Function__
<font color='green'>__Access to Industry Index (INDI) Historical Data (= شاخص صنعت بورس)__</font>

This is a cap-weighted index for just those stocks that are industrial companies. This function gets the industry index (INDI) data for you. Please note that the historical data for MKT2I index, in the form of OHLCV, is available from 1387-09-15, but adj-close historical data is available from 1387-09-14. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [15]:
indi_history = fpy.Get_INDI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
indi_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    66988.0    67547.0    66988.0    67547.0   
1395-01-08 2016-03-27     Sunday    67595.0    67607.0    67493.0    67493.0   
1395-01-09 2016-03-28     Monday    67276.0    67314.0    67209.0    67209.0   
1395-01-10 2016-03-29    Tuesday    67166.0    67166.0    66874.0    66875.0   
1395-01-11 2016-03-30  Wednesday    67102.0    67589.0    67102.0    67589.0   
...               ...        ...        ...        ...        ...        ...   
1400-11-03 2022-01-23     Sunday  1178980.0  1178980.0  1158210.0  1158210.0   
1400-11-04 2022-01-24     Monday  1153620.0  1153620.0  1135820.0  1135820.0   
1400-11-05 2022-01-25    Tuesday  1132650.0  1132650.0  1109490.0  1109490.0   
1400-11-06 2022-01-26  Wednesday  1122510.0  1131570.0  1122510.0  1131570.0   
1400-11-09 2022-01-29   Saturday  1140780.0  1161190.0  1140780.0  1161170.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    67547.1   571954728  
1395-01-08    67492.9   529072524  
1395-01-09    67209.4   397361929  
1395-01-10    66874.7  1151654864  
1395-01-11    67588.8   681107453  
...               ...         ...  
1400-11-03  1158190.0  3551904061  
1400-11-04  1135840.0  2480333502  
1400-11-05  1109420.0  3988836441  
1400-11-06  1131520.0  3672607090  
1400-11-09  1161190.0  2775624703  

[1414 rows x 8 columns]

### __3.9. Get_LCI30_History() Function__
<font color='green'>__Access to Top 30 Large-Cap Index (LCI30) Historical Data (= شاخص 30 شرکت بزرگ بورس)__</font>

This is a cap-weighted index for top 30 large-cap (LCI30) stocks in Bourse market. This function gets the 30 large-cap index (LCI30) data for you. Please note that the historical data for 30LCI index, in the form of OHLCV and adj-close, both are available from 1389-05-24. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [17]:
lci30_history = fpy.Get_LCI30_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
lci30_history

Date    Weekday     Open     High      Low    Close  \
J-Date                                                                 
1395-01-07 2016-03-26   Saturday   3379.0   3454.0   3379.0   3454.0   
1395-01-08 2016-03-27     Sunday   3465.0   3475.0   3455.0   3455.0   
1395-01-09 2016-03-28     Monday   3444.0   3444.0   3431.0   3431.0   
1395-01-10 2016-03-29    Tuesday   3421.0   3423.0   3401.0   3401.0   
1395-01-11 2016-03-30  Wednesday   3404.0   3452.0   3404.0   3450.0   
...               ...        ...      ...      ...      ...      ...   
1400-11-03 2022-01-23     Sunday  70087.0  70087.0  68533.0  68533.0   
1400-11-04 2022-01-24     Monday  68504.0  68504.0  67189.0  67189.0   
1400-11-05 2022-01-25    Tuesday  67182.0  67182.0  65808.0  65812.0   
1400-11-06 2022-01-26  Wednesday  65981.0  67725.0  65981.0  67725.0   
1400-11-09 2022-01-29   Saturday  67859.0  69836.0  67859.0  69836.0   

            Adj Close      Volume  
J-Date                             
1395-01-07     3454.2   290228382  
1395-01-08     3455.4   336982051  
1395-01-09     3430.6   255295732  
1395-01-10     3400.7   411464948  
1395-01-11     3450.4   323150383  
...               ...         ...  
1400-11-03    68528.4  1617170488  
1400-11-04    67189.1  1055882555  
1400-11-05    65806.9  2198118902  
1400-11-06    67719.9  1497567879  
1400-11-09    69840.5   805017064  

[1414 rows x 8 columns]

### __3.10. Get_ACT50_History() Function__
<font color='green'>__Access to Top 50 Active Index (ACT50) Historical Data (= شاخص 50 شرکت فعال بورس)__</font>

This is a cap-weighted index for top 50 active (ACT50) stocks in Bourse market. This function gets the 50 most active stocks index (ACT50) data for you. Please note that the historical data for ACT50 index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-15. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1.

In [18]:
act50_history = fpy.Get_ACT50_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
act50_history

Date    Weekday     Open     High      Low    Close  \
J-Date                                                                 
1395-01-07 2016-03-26   Saturday   3256.0   3301.0   3256.0   3301.0   
1395-01-08 2016-03-27     Sunday   3311.0   3311.0   3305.0   3305.0   
1395-01-09 2016-03-28     Monday   3293.0   3296.0   3290.0   3290.0   
1395-01-10 2016-03-29    Tuesday   3287.0   3287.0   3271.0   3271.0   
1395-01-11 2016-03-30  Wednesday   3283.0   3311.0   3283.0   3311.0   
...               ...        ...      ...      ...      ...      ...   
1400-11-03 2022-01-23     Sunday  50842.0  50842.0  49915.0  49915.0   
1400-11-04 2022-01-24     Monday  49706.0  49706.0  48929.0  48929.0   
1400-11-05 2022-01-25    Tuesday  48789.0  48789.0  47839.0  47839.0   
1400-11-06 2022-01-26  Wednesday  48451.0  48883.0  48451.0  48883.0   
1400-11-09 2022-01-29   Saturday  49307.0  50172.0  49307.0  50172.0   

            Adj Close      Volume  
J-Date                             
1395-01-07     3300.7   631986992  
1395-01-08     3304.8   775084867  
1395-01-09     3289.8   470238679  
1395-01-10     3270.6  1127454520  
1395-01-11     3310.8   644347187  
...               ...         ...  
1400-11-03    49912.9  2404775518  
1400-11-04    48929.5  1747070509  
1400-11-05    47836.0  3236305736  
1400-11-06    48880.5  2312464979  
1400-11-09    50173.4  1443993005  

[1414 rows x 8 columns]

### __3.11. Get_SectorIndex_History() Function__
<font color='green'>__Access to a Sector Index Historical Data (= شاخص گروه صنعت در بازار بورس)__</font>

Each sector has its own cap-weighted index. The universe of each sector index is that sector's stocks in the Bourse (not Fara-Bourse) market. You can access to each sector cap-weighted index using this function and sector name. 

The inputs and outputs are same as __Get_CWI_History()__ described in section 3.1, except that you need to enter a __Farsi sector index name__ as __sector__ input.

In [19]:
df_sector_index_history = fpy.Get_SectorIndex_History(sector = ' بانکیها', start_date='1395-01-01', end_date='1400-12-29', 
                                   ignore_date = False, just_adj_close = False, show_weekday = True, double_date = True)
df_sector_index_history

Date    Weekday    Open    High     Low   Close  Adj Close  \
J-Date                                                                        
1395-01-07 2016-03-26   Saturday   753.0   781.0   753.0   781.0      780.9   
1395-01-08 2016-03-27     Sunday   792.0   792.0   787.0   789.0      789.1   
1395-01-09 2016-03-28     Monday   788.0   789.0   788.0   788.0      787.5   
1395-01-10 2016-03-29    Tuesday   787.0   787.0   784.0   784.0      784.0   
1395-01-11 2016-03-30  Wednesday   786.0   796.0   786.0   795.0      794.9   
...               ...        ...     ...     ...     ...     ...        ...   
1400-11-03 2022-01-23     Sunday  5713.0  5713.0  5607.0  5607.0     5605.9   
1400-11-04 2022-01-24     Monday  5604.0  5604.0  5586.0  5586.0     5585.8   
1400-11-05 2022-01-25    Tuesday  5589.0  5589.0  5506.0  5506.0     5506.4   
1400-11-06 2022-01-26  Wednesday  5511.0  5532.0  5511.0  5532.0     5532.4   
1400-11-09 2022-01-29   Saturday  5541.0  5597.0  5541.0  5597.0     5598.1   

               Volume  
J-Date                 
1395-01-07  134013625  
1395-01-08  191566403  
1395-01-09  161292181  
1395-01-10  176754056  
1395-01-11  161472514  
...               ...  
1400-11-03  339939904  
1400-11-04  337781553  
1400-11-05  681398783  
1400-11-06  230021331  
1400-11-09  137151351  

[1414 rows x 8 columns]

---------------------------------------------
## __4. Reading Real-Time Market Data__
<font color='green'>__Gathering Real-Time Data of Marketwatch when the Market is Open (= جمع آوری اطلاعات لحظه ای مربوط به دفترچه سفارشات و مارکت واچ همه سهم ها)__</font>

__Important Note:__ OB stands for Order Book or in Farsi دفتر سفارشات.


Sometimes, traders make their decisions based on the market real-time data, when the market is open. For example you may have an strategy based on the real-time orderbook data or clients (retail and institutional) data. If this is the case for you, you will need this function. As the function returns order book data, it is also necessary for optimal trade execution. 

The function inputs are:


| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | :- | -: |
| __save_excel__ | True/Flase | If True, fucntion saves data in the given path as Excel file (2 Excel Files) |ذخیره یا عدم ذخیره دیتا فریم های خروجی (دو دیتا فریم) به صورت فایل اکسل |
| __save_path__ | String | Path on your local drive to save the gathered data |مسیر مورد نظر برای ذخیره دیتا فریم های خروجی (دو دیتا فریم)|

The function will return two dataframes, the first one is the __market watch real-time data__ and the second is the order book real-time data. The first dataframe have the following columns:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __Ticker (index)__ |Ticker| نماد |
| __Trade Type__ |Trade Type| نوع معامله: تابلو، بلوکی، عمده، جبرانی |
| __Time__ |Time| زمان مربوط به اطلاعات نمایش داده شده یک نماد |
| __Open__ |Day Open Price| قیمت باز شدن روز جاری |
| __High__ |Day High Price| بالاترین قیمت رخ داده در روز جاری تا آن لحظه |
| __Low__ |Day Low Price| پایین ترین قیمت رخ داده در روز جاری تا آن لحظه |
| __Close__ |Day Close Price| آخرین قیمت معامله در آن لحظه |
| __Final__ |Day Final Price| قیمت پایانی در آن لحظه |
| __Close(%)__ |Day Close % Change with respect to yesterday Final Price| درصد تغییر آخرین قیمت معامله در آن لحظه نسبت به قیمت پایانی دیروز |
| __Final(%)__ |Day Final % Change with respect to yesterday Final Price| درصد تغییر قیمت پایانی در آن لحظه نسبت به قیمت پایانی دیروز |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز روز جاری |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز روز جاری |
| __Value__ |Value of Executed Trades| ارزش معاملات در روز جاری تا آن لحظه |
| __BQ-Value__ |Buy Queue Value| ارزش صف خرید در نماد مورد نظر در آن لحظه |
| __SQ-Value__ |Sell Queue Value| ارزش صف فروش در نماد مورد نظر در آن لحظه |
| __BQPC__ |Buy Queue Per-Capita (Average)| ارزش سرانه افرادی که در آن لحظه در صف خرید نماد مورد نظر هستند |
| __SQPC__ |Sell Queue Per-Capita (Average)| ارزش سرانه افرادی که در آن لحظه در صف فروش نماد مورد نظر هستند |
| __Volume__ |Volume of Executed Trades| حجم معاملات در نماد مورد نظر تا آن لحظه |
| __Vol_Buy_R__ |Retail Traders Buy Volume| حجم خرید افراد حقیقی در نماد مورد نظر تا آن لحظه |
| __Vol_Buy_I__ |Institutioanl Traders Buy Volume| حجم خرید افراد حقوقی در نماد مورد نظر تا آن لحظه |
| __Vol_Sell_R__ |Retail Traders Sell Volume| حجم فروش افراد حقیقی در نماد مورد نظر تا آن لحظه |
| __Vol_Sell_I__ |Institutional Traders Sell Volume| حجم فروش افراد حقوقی در نماد مورد نظر تا آن لحظه |
| __No__ |Number of Executed Trades| تعداد معاملات انجام شده در نماد مورد نظر تا آن لحظه |
| __No_Buy_R__ |Number of Buyer Retail Traders| تعداد خریداران حقیقی در نماد مورد نظر تا آن لحظه |
| __No_Sell_R__ |Number of Seller Retail Traders| تعداد فروشندگان حقیقی در نماد مورد نظر تا آن لحظه |
| __No_Buy_I__ |Number of Buyer Institutional Retail Traders| تعداد خریداران حقوقی در نماد مورد نظر تا آن لحظه |
| __No_Sell_I__ |Number of Seller Institutional Traders| تعداد فروشندگان حقوقی در نماد مورد نظر تا آن لحظه |
| __Name__ |Firm Farsi Full Name| نام فارسی شرکت مربوط به نماد مورد نظر |
| __Market__ |Category and Market of the Ticker| بازار نماد مورد نظر: بورس، فرابورس، پایه، حق تقدم بورس، حق تقدم فرابورس، حق تقدم پایه و صندوق های قابل معامله |
| __Sector__ |Sector/Industry of the ticker| گروه صنعت نماد مورد نظر |
| __Share-No__ |Number of Released Shares of the Ticker| تعداد سهام نماد مورد نظر |
| __Base-Vol__ |Base Volume of the Ticker| حجم مبنای نماد مورد نظر |
| __Market Cap__ |Market Cap of the Ticker (based on Final Price)| ارزش بازار نماد مورد نظر در آن لحظه - محاسبه شده بر اساس قیمت پایانی|
| __EPS__ | TTM EPS (TSE Data) | نماد مورد نظر به صورت میانگین دوازده ماهه متحرک EPS|
| __Download__ |Download Jalali Date and Time| تاریخ شمسی و زمان دانلود اطلاعات |

The second dataframe is for __real-time order book data__ and have the following columns:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __Ticker (index)__ |Ticker| نماد |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز برای معامله در روز جاری |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز برای معامله در روز جاری |
| __OB-Depth__ |Order Book Depth (1-5)| عمق دفتر سفارشات (بهترین مظنه در عمق 1) |
| __Sell-No__ |Number of People Willing to Sell at Given Price| تعداد افراد حاضر به فروش در آن مظنه فروش |
| __Sell-Vol__ |Day Open Price| حجم تجمعی فروشندگان در آن مظنه فروش |
| __Sell-Price__ |Offered Sell Price| قیمت پیشنهادی فروشندگان در آن مظنه فروش |
| __Buy-Price__ |Offered Buy Price| قیمت پیشنهادی خریداران در آن مظنه خرید |
| __Buy-Vol__ |Ready Buy Volume at Given Price| حجم تجمعی خریداران در آن مظنه خرید  |
| __Buy-No__ |Number of People Willing to Buy at Given Price| تعداد افراد حاضر به خرید در آن مظنه خرید |
| __Download__ |Download Jalali Date and Time| تاریخ شمسی و زمان دانلود اطلاعات |

Both dataframes will be returned by this function and you can manage them in your desired databases if you will. If you want the dataframes to be saved as excel in your local drive, you need to say this to the function and specify a path for saving data using the following input parameters:
```python  
save_excel = True
save_path = 'D:/FinPy-TSE Data/MarketWatch'
```
The Excel file names, will be __date and time stamped__, like __MarketWatch 1400-10-10 12-02-29__ and __OrderBook 1400-10-10 12-02-29__. So you can use them later or save them in your desired databases, for future analysis.

In [20]:
df_market_watch, df_order_book = fpy.Get_MarketWatch(save_excel = True, save_path = 'D:/FinPy-TSE Data/MarketWatch')
df_market_watch

Trade Type      Time     Open     High      Low    Close    Final  \
Ticker                                                                     
وسپهر       تابلو  12:29:59   7060.0   7250.0   7060.0   7170.0   7080.0   
شصدف        تابلو  12:29:07  32000.0  32200.0  31800.0  32200.0  31450.0   
فسا         تابلو  12:29:55    847.0    847.0    847.0    847.0    860.0   
فرآور       تابلو  12:28:56  22640.0  22640.0  22640.0  22640.0  22490.0   
سبزوا       تابلو  12:29:26   1522.0   1580.0   1520.0   1580.0   1545.0   
...           ...       ...      ...      ...      ...      ...      ...   
ولصنم       تابلو  12:29:10   3330.0   3449.0   3158.0   3449.0   3417.0   
غصینو       تابلو  12:29:59  14950.0  15690.0  14520.0  15280.0  15180.0   
کارا        تابلو  14:56:24  10869.0  10869.0  10656.0  10656.0  10657.0   
سپ          تابلو  12:23:48  21340.0  21340.0  21340.0  21340.0  20600.0   
شپترو       تابلو  12:24:14   2310.0   2376.0   2310.0   2376.0   2373.0   

        Close(%)  Final(%)   Day_UL  ...  No_Sell_R  No_Sell_I  \
Ticker                               ...                         
وسپهر       1.56      0.28   7410.0  ...     1436.0        1.0   
شصدف        4.89      2.44  32200.0  ...      386.0        0.0   
فسا        -2.98     -1.49    899.0  ...       45.0        2.0   
فرآور       4.96      4.27  22640.0  ...       55.0        1.0   
سبزوا       4.98      2.66   1580.0  ...      154.0        1.0   
...          ...       ...      ...  ...        ...        ...   
ولصنم       4.99      4.02   3449.0  ...      124.0        1.0   
غصینو       2.00      1.34  15720.0  ...      335.0        3.0   
کارا        0.01      0.02  11720.0  ...      118.0        5.0   
سپ          4.97      1.33  21340.0  ...       40.0        3.0   
شپترو       2.99      2.86   2376.0  ...      124.0        0.0   

                                Name             Market  \
Ticker                                                    
وسپهر   سرمایه گذاری مالی سپهرصادرات            فرابورس   
شصدف                  صنعتی دوده فام            فرابورس   
فسا                     پتروشیمی فسا               پایه   
فرآور        فرآوری موادمعدنی ایران                بورس   
سبزوا               سیمان لار سبزوار            فرابورس   
...                              ...                ...   
ولصنم             لیزینگ صنعت ومعدن                بورس   
غصینو                     صنعتی مینو            فرابورس   
کارا                صندوق س. کارا -د  صندوق قابل معامله   
سپ        پرداخت الکترونیک سامان کیش               بورس   
شپترو                پتروشیمی آبادان               پایه   

                                          Sector      Share-No    Base-Vol  \
Ticker                                                                       
وسپهر                             سرمایه گذاریها  3.000000e+10  12000000.0   
شصدف                             محصولات شیمیایی  7.000000e+08    488600.0   
فسا                              محصولات شیمیایی  3.486000e+09  11454754.0   
فرآور                                فلزات اساسی  6.600000e+08    695410.0   
سبزوا                            سیمان، آهک و گچ  3.757798e+09   9966778.0   
...                                          ...           ...         ...   
ولصنم                     سایر واسطه گریهای مالی  2.000000e+09   4566210.0   
غصینو   محصولات غذایی و آشامیدنی به جز قند و شکر  1.000000e+10   4000000.0   
کارا              صندوق سرمایه گذاری قابل معامله  1.000000e+08         1.0   
سپ              رایانه و فعالیت های وابسته به آن  8.000000e+09   3200000.0   
شپترو                            محصولات شیمیایی  5.520000e+09   4334634.0   

          Market Cap     EPS             Download  
Ticker                                             
وسپهر   2.124000e+14   631.0  1400-11-10 01:10:48  
شصدف    2.201500e+13  2314.0  1400-11-10 01:10:48  
فسا     2.997960e+12     2.0  1400-11-10 01:10:48  
فرآور   1.484340e+13  1576.0  1400-11-10 01:10:48  
سبزوا   5.805797e+12   295.0  1400-11-10 01:10:48  
...         

In [21]:
df_order_book.head(15)

Sell-No  Sell-Vol  Sell-Price  Buy-Price  \
Ticker Day_LL Day_UL OB-Depth                                             
آ س پ  4400   4862   1               2      6001        4640       4520   
                     2               1      8000        4642       4501   
                     3               3     20040        4650       4484   
                     4               2     10223        4669       4464   
                     5               1         1        4680       4463   
آبادا  8320   9180   1               3      2498        9070       9050   
                     2               2     10048        9080       9040   
                     3               2       948        9090       9030   
                     4               6     19285        9100       9020   
                     5               5       240        9120       9010   
آبین   1428   1516   1               1       100        1499       1472   
                     2               1     10000        1505       1470   
                     3               1     10000        1509       1466   
                     4               2    111882        1510       1460   
                     5               1     35605        1513       1457   

                               Buy-Vol  Buy-No             Download  
Ticker Day_LL Day_UL OB-Depth                                        
آ س پ  4400   4862   1           20000       1  1400-11-10 01:10:48  
                     2           60000       1  1400-11-10 01:10:48  
                     3            8875       2  1400-11-10 01:10:48  
                     4            2000       1  1400-11-10 01:10:48  
                     5            2271       1  1400-11-10 01:10:48  
آبادا  8320   9180   1            1111       1  1400-11-10 01:10:48  
                     2             642       1  1400-11-10 01:10:48  
                     3            6286       1  1400-11-10 01:10:48  
                     4           24359       4  1400-11-10 01:10:48  
                     5            2650       1  1400-11-10 01:10:48  
آبین   1428   1516   1           11281       2  1400-11-10 01:10:48  
                     2            5000       1  1400-11-10 01:10:48  
                     3           10000       1  1400-11-10 01:10:48  
                     4           48118       1  1400-11-10 01:10:48  
                     5           70000       1  1400-11-10 01:10:48

---------------------------------------------
## __5. Collective Data Download__

### __5.1. Build_Market_StockList() Function__
<font color='green'>__Gathering Detailed Information of All Stocks, Available in TSE (= تهیه لیست سهم های بازار بورس، فرابورس و پایه)__</font>

Every investor or trader needs to know all of the options available for investment or trading in a given market. This function, provides you a list of all stocks from __Bourse__, __Fara-Bourse__ and __Payeh__ market. It would be a good practice to save the output of this function in your local drive, play around with it in Python (or even in Excel, using filter functionality), to choose a list of stocks that meets your requirement. 

The function inputs are:


| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | :- | -: |
| __bourse__ |True/Flase | If True, the fucntion considers Bourse stocks data in the output list |احتساب یا عدم احتساب سهم های بازار بورس در جمع آوری اطلاعات و تهیه لیست خروجی |
| __farabourse__ | True/Flase | If True, the fucntion considers Fara-Bourse stocks data in the output list |احتساب یا عدم احتساب سهم های بازار فرابورس در جمع آوری اطلاعات و تهیه لیست خروجی |
| __payeh__ | True/Flase | If True, the fucntion considers Payeh stocks data in the output list |احتساب یا عدم احتساب سهم های بازار پایه در جمع آوری اطلاعات و تهیه لیست خروجی |
| __detailed_list__ | True/Flase | If False, just returns you the Ticker, Name and Market columns, otherwise it shows the remaining columns|جمع آوری یا عدم جمع آوری اطلاعات تکمیلی سهم ها|
| __show_progress__ | True/Flase | If True, shows % progress of gathering and processing data |نمایش یا عدم نمایش میزان پیشرفت در جمع آوری اطلاعات |
| __save_excel__ | True/Flase | If True, fucntion saves data in the given path as Excel file |ذخیره یا عدم ذخیره لیست تهیه شده خروجی به صورت فایل اکسل |
| __save_csv__ | True/Flase | If True, fucntion saves data in the given path as CSV file |CSV ذخیره یا عدم ذخیره لیست تهیه شده خروجی به صورت فایل |
| __save_path__ | String | Path on your local drive to save the gathered data |مسیر مورد نظر برای ذخیره لیست تهیه شده|

The output will be a list of stocks with __Ticker__ as index and __Name, Market, Panel, Sector, Sub-Sector, Comment, Name(EN), Company Code(12), Ticker(4), Ticker(5), Ticker(12), Sector Code, Sub-Sector Code and Panel Code__, depending on if you want detailed list or not.

In [22]:
df_stock_list = fpy.Build_Market_StockList(bourse = True, farabourse = True, payeh = True, detailed_list = True, show_progress = True, 
                                           save_excel = True, save_csv = False, save_path = 'D:/FinPy-TSE Data/Stock List')
df_stock_list 

Progress : 100 % , Done in 19 seconds!
File saved in the specificed directory as:  1400-11-10 detailed_stocklist_bfp.xlsx


Name       Market  \
Ticker                                             
آبادا        تولید نیروی برق آبادان         بورس   
آپ                آسان پرداخت پرشین         بورس   
آسیا                      بیمه آسیا         بورس   
آکنتور             کنتورسازی ایران          بورس   
اپال    فرآوری معدنی اپال کانی پارس         بورس   
...                             ...          ...   
توسعه                 اعتباری توسعه    پایه قرمز   
فاهواز            نورد و لوله اهواز  پایه نارنجی   
کیسون                    شرکت کیسون     پایه زرد   
سایرا                      ایرانیت   پایه نارنجی   
سامان                    بانک سامان     پایه زرد   

                               Panel  \
Ticker                                 
آبادا                 بازار دوم بورس   
آپ                    بازار دوم بورس   
آسیا    بازار اول (تابلوی اصلی) بورس   
آکنتور  بازار اول (تابلوی فرعی) بورس   
اپال                  بازار دوم بورس   
...                              ...   
توسعه        بازار پایه قرمز فرابورس   
فاهواز     بازار پایه نارنجی فرابورس   
کیسون         بازار پایه زرد فرابورس   
سایرا      بازار پایه نارنجی فرابورس   
سامان         بازار پایه زرد فرابورس   

                                          Sector  \
Ticker                                             
آبادا                 عرضه برق، گاز، بخاروآب گرم   
آپ              رایانه و فعالیت های وابسته به آن   
آسیا    بیمه وصندوق بازنشستگی به جزتامین اجتماعی   
آکنتور          ابزارپزشکی، اپتیکی و اندازه گیری   
اپال                       استخراج کانه های فلزی   
...                                          ...   
توسعه                    بانکها و موسسات اعتباری   
فاهواز                               فلزات اساسی   
کیسون                انبوه سازی، املاک و مستغلات   
سایرا                  سایر محصولات کانی غیرفلزی   
سامان                    بانکها و موسسات اعتباری   

                                  Sub-Sector  \
Ticker                                         
آبادا         تولید انتقال و توزیع نیروی برق   
آپ                   مشاوره و تهیه نرم افزار   
آسیا                           بیمه غیرزندگی   
آکنتور  وسایل اندازه گیری غیر از کنترل صنعتی   
اپال                             استخراج آهن   
...                                      ...   
توسعه                    سایر واسطه های پولی   
فاهواز                تولید آهن و فولاد پایه   
کیسون              پیمانکاری املاک و مستغلات   
سایرا        تولید محصولات بتنی، سیمانی و گچ   
سامان                    سایر واسطه های پولی   

                                                  Comment            Name(EN)  \
Ticker                                                                          
آبادا                                                   -           Abadan PG   
آپ                                                      -  Asan Pardakht Pers   
آسیا                                                    -           Asia Bime   
آکنتور                                                  -        Iran Counter   
اپال                                                    -      Opal Kani Pars   
...                                                   ...                 ...   
توسعه    قانون احکام دائمی برنامه های توسعه کشور/ اظها...      Etebari Dev Co   
فاهواز                                      لغو پذیرش شده    Ahwaz Rolling Co   
کیسون    مشمول ماده 36 قانون احکام دائمی برنامه هاي تو...            S*Kayson   
سایرا                       لغو پذیرش شده/ اظهارنظر مردود              Iranit   
سامان             قانون احکام دائمی برنامه های توسعه کشور        S*Saman Bank   

       Company Code(12) Ticker(4) Ticker(5)    Ticker(12) Sector Code  \
Ticker                                                                  
آبادا      IRO1NBAB0009      NBAB     NBAB1  IRO1NBAB0001          40   
آپ         IRO1APPE0008      APPE     APPE1  IRO1APPE0001          72   
آسیا       IRO1ASIA0008      ASIA     ASIA1  IRO1ASIA0001          66   
آکنتور     IRO1CONT0002      CONT     CONT1  IRO1CONT0001          33   
اپال       IRO1OPAL0000      OPAL     O

### __5.2. Build_PricePanel() Function__
<font color='green'>__Buliding Price Panel for Given List of Stocks (= پنل قیمت برای لیست دلخواه از سهم ها بر اساس قیمت پایانی تعدیل شده)__</font>

In stock selection, portfolio management and some other applications, we need to have some sort tabular data, to be able to feed our algorithm. For example, in the following table, you can see __Adj Final__ price data of 10 stocks for 1000 days. Columns of this table are stock tickers, with dates as index. Table values are Adj Final price of each stock for that date. The best practice is to have this table in the form of a Pandas data frame, which is the way we have implemented this function.

| __Date__ | __Stock 1__ | __Stock 2__ | __Stock 3__ |__Stock 4__ |__Stock 5__ | __Stock 6__ | __Stock 7__ | __Stock 8__ |__Stock 9__ |__Stock 10__ |
| :- | :-: | :-: |:-: | :-: | :-: |:-: | :-: |:-: | :-: | :-: | 
| __Date 1__ | $P_{1,1}$ |$P_{2,1}$ |$P_{3,1}$ |$P_{4,1}$ |$P_{5,1}$ |$P_{6,1}$ |$P_{7,1}$ |$P_{8,1}$ |$P_{9,1}$ |$P_{10,1}$ |
| __Date 2__ | $P_{1,2}$ |$P_{2,2}$ |$P_{3,2}$ |$P_{4,2}$ |$P_{5,2}$ |$P_{6,2}$ |$P_{7,2}$ |$P_{8,2}$ |$P_{9,2}$ |$P_{10,2}$ |
| __Date 3__ | $P_{1,3}$ |$P_{2,3}$ |$P_{3,3}$ |$P_{4,3}$ |$P_{5,3}$ |$P_{6,3}$ |$P_{7,3}$ |$P_{8,3}$ |$P_{9,3}$ |$P_{10,3}$ |
| __...__ | ... |... |... |... |... |... |... |... |... |... |
| __Date 500__ | $P_{1,500}$ |$P_{2,500}$ |$P_{3,500}$ |$P_{4,500}$ |$P_{5,500}$ |$P_{6,500}$ |$P_{7,500}$ |$P_{8,500}$ |$P_{9,500}$ |$P_{10,500}$ |
| __Date 501__ | $P_{1,501}$ |$P_{2,501}$ |$P_{3,501}$ |$P_{4,501}$ |$P_{5,501}$ |$P_{6,501}$ |$P_{7,501}$ |$P_{8,501}$ |$P_{9,501}$ |$P_{10,501}$ |
| __Date 502__ | $P_{1,502}$ |$P_{2,502}$ |$P_{3,502}$ |$P_{4,502}$ |$P_{5,502}$ |$P_{6,502}$ |$P_{7,502}$ |$P_{8,502}$ |$P_{9,502}$ |$P_{10,502}$ |
| __...__ | ... |... |... |... |... |... |... |... |... |... |
| __Date 998__ | $P_{1,998}$ |$P_{2,998}$ |$P_{3,998}$ |$P_{4,998}$ |$P_{5,998}$ |$P_{6,998}$ |$P_{7,998}$ |$P_{8,998}$ |$P_{9,998}$ |$P_{10,998}$ |
| __Date 999__ | $P_{1,999}$ |$P_{2,999}$ |$P_{3,999}$ |$P_{4,999}$ |$P_{5,999}$ |$P_{6,999}$ |$P_{7,999}$ |$P_{8,999}$ |$P_{9,999}$ |$P_{10,999}$ |
| __Date 1000__ | $P_{1,1000}$ |$P_{2,1000}$ |$P_{3,1000}$ |$P_{4,1000}$ |$P_{5,1000}$ |$P_{6,1000}$ |$P_{7,1000}$ |$P_{8,1000}$ |$P_{9,1000}$ |$P_{10,1000}$ |

The function inputs are:


| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | :- | -: |
| __stock_list__ |Python List | List of stocks (tickers or firm names) |لیستی از نماد یا نام شرکت ها |
| __jalali_date__ | True/Flase | If True, fucntion returns Jalali date as index, otherwise it will return iternational date |تعیین تاریخ شمسی یا میلادی برای ایندکس پنل قیمت |
| __save_excel__ | True/Flase | If True, fucntion saves individual stocks data and price panel, in the given path |ذخیره یا عدم ذخیره اطلاعات دانلود شده برای تک تک سهم ها و پنل قیمت |
| __save_path__ | String | Path on your local drive to save the downloaded data |مسیر ذخیره اطلاعات تک تک سهم ها و پنل قیمت|

The function output will be a dataframe like what you see above, with tickers as column names (irrespective of if you entered stock names or tickers in the input list) and __J-Date__ or __Date__ as index, depending your input settings. The function also saves the downloaded data as Excel file, if you use proper input settings.

__Important Note:__ The speed of gathering and processing data, depends on your internet speed, TSE servers response speed and if you want to save data as Excel or not. A rule of thumb for estimating runing time of the function is _1sec/stock_ if you do not save Excel files and _2sec/stock_ if you want to save the data. For example if you have 60 stocks and you do not want to save the data as Excel, it will take about 60 seconds (1 minute) to gather and process the data. Remember, if you run this function more than usual, the TSE servers will block your IP and you need to wait for a few minutes to be able to get data from TSE, again.

In [3]:
# create a list of stocks using previous function output:
bourse_farabourse_stocks = df_stock_list[(df_stock_list['Market']=='بورس')|((df_stock_list['Market']=='فرابورس'))]
bourse_farabourse_stocks = bourse_farabourse_stocks.index.tolist()    # extract tickers and convert to list
# now call the price panel function and pass the list to it:
df_price_panel = fpy.Build_PricePanel(stock_list = bourse_farabourse_stocks, jalali_date = True, 
                                      save_excel = True, save_path = 'D:/FinPy-TSE Data/Price Panel')
df_price_panel

Reading Data :  100.0 %    Processing and Saving Data :  100.0 %
1164 Seconds Took to Gather and Process Your Requested Data


آبادا    آسیا      آپ  آکنتور  اتکام   اخابر  افق   البرز  \
J-Date                                                                   
1380-01-05     NaN     NaN     NaN     NaN    NaN     NaN  NaN     NaN   
1380-01-06     NaN     NaN     NaN     NaN    NaN     NaN  NaN     NaN   
1380-01-07     NaN     NaN     NaN     NaN    NaN     NaN  NaN     NaN   
1380-01-08     NaN     NaN     NaN     NaN    NaN     NaN  NaN     NaN   
1380-01-11     NaN     NaN     NaN     NaN    NaN     NaN  NaN     NaN   
...            ...     ...     ...     ...    ...     ...  ...     ...   
1400-11-02  8940.0  1504.0  7630.0     NaN    NaN  6830.0  NaN  1601.0   
1400-11-03  8860.0  1430.0  7270.0     NaN    NaN  6640.0  NaN  1544.0   
1400-11-04  8810.0  1364.0  6920.0     NaN    NaN  6590.0  NaN  1474.0   
1400-11-05  8750.0  1300.0  6620.0     NaN    NaN  6460.0  NaN  1407.0   
1400-11-06  8750.0  1330.0  6420.0     NaN    NaN  6460.0  NaN  1436.0   

              امید     امین  ...    کیمیا     گدنا  گوهران   گکوثر   ولشرق  \
J-Date                       ...                                             
1380-01-05     NaN      NaN  ...      NaN      NaN     NaN     NaN     NaN   
1380-01-06     NaN      NaN  ...      NaN      NaN     NaN     NaN     NaN   
1380-01-07     NaN      NaN  ...      NaN      NaN     NaN     NaN     NaN   
1380-01-08     NaN      NaN  ...      NaN      NaN     NaN     NaN     NaN   
1380-01-11     NaN      NaN  ...      NaN      NaN     NaN     NaN     NaN   
...            ...      ...  ...      ...      ...     ...     ...     ...   
1400-11-02     NaN  10210.0  ...  11110.0  11390.0  4976.0  1300.0  1976.0   
1400-11-03     NaN  10220.0  ...  10950.0  10940.0  4977.0  1251.0  1966.0   
1400-11-04     NaN  10220.0  ...  10480.0  10700.0  4951.0  1243.0  1943.0   
1400-11-05  5940.0  10240.0  ...  10110.0  10280.0  4925.0  1196.0  1920.0   
1400-11-06  5940.0  10240.0  ...   9850.0  10210.0  4956.0  1243.0  1919.0   

             ومعلم    وملل   ومهان  وهامون    وهور  
J-Date                                              
1380-01-05     NaN     NaN     NaN     NaN     NaN  
1380-01-06     NaN     NaN     NaN     NaN     NaN  
1380-01-07     NaN     NaN     NaN     NaN     NaN  
1380-01-08     NaN     NaN     NaN     NaN     NaN  
1380-01-11     NaN     NaN     NaN     NaN     NaN  
...            ...     ...     ...     ...     ...  
1400-11-02  1732.0  9340.0  4018.0     NaN  1105.0  
1400-11-03  1708.0  9240.0  3999.0     NaN  1134.0  
1400-11-04  1676.0  9160.0  3990.0     NaN  1161.0  
1400-11-05  1638.0  9040.0  3926.0     NaN  1139.0  
1400-11-06  1621.0  9010.0  3939.0     NaN  1149.0  

[5035 rows x 531 columns]

---------------------------------------------
## __6. Reading USD/Rial Exchange Rate__

Although USD/Rial exchange rate is out of the scope of Tehran Stock Exchange, but as it has important effect on the price of stocks, we decided to include a function in finpy-tse module, to access this data, as well. 

### __6.1. Get_USD-RIAL() Function__
<font color='green'>__Access to USD/RIAL Exchange Rate Historical Data (= نرخ دلار آمریکا در بازار آزاد)__</font>

The function has similar inputs to our other functions. It returns daily price of USD/Rial and only supports last 10-years of daily history.


| __Parameter__ | __Format__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | :- | -: |
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date |تاریخ شروع (شمسی) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date |تاریخ پایان (شمسی) |
| __ignore_date__ | True/Flase | if True, fucntion does not consider start and end date and gets all available history |نادیده گرفتن تاریخ شروع و پایان و نمایش همه سابقه در دسترس |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |نمایش یا عدم نمایش روزهای هفته در خروجی |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |نمایش یا عدم نمایش تاریخ میلادی (علاوه بر تاریخ شمسی) |

The output of the function will be a pandas dataframe with the follwing columns:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ |
| :- | :- | -: | 
| __J-Date__ | Jalali Date (as dataframe index) | تاریخ شمسی (ایندکس دیتا فریم خروجی) |
| __Date__ | International Date | تاریخ میلادی |
| __Weekday__ | Name of weekdays | نام روزهای هفته |
| __Open__ | Daily open price | اولین قیمت روز |
| __High__ | Daily high price | بالاترین قیمت روز |
| __Low__ | Daily low price | پایین ترین قیمت روز |
| __Close__ | Daily close price | آخرین قیمت روز |

By using proper inputs, when calling the function, you can drop __Date__ and __Weekday__ columns, if you will.

__Important Note:__ The reference of USD/Rial exchange rate is https://www.tgju.org/profile/price_dollar_rl/history. Sometimes the rates announced by this website, is different from the real prices, because the policies of Central Bank of Iran (CBI) for announcement of rates changes from time to time.

In [23]:
df_usd = fpy.Get_USD_RIAL(start_date = '1395-01-01', end_date = '1400-12-29',\
                          ignore_date = False, show_weekday = True, double_date = True)
df_usd

Date    Weekday    Open    High     Low   Close
J-Date                                                          
1395-01-07 2016-03-26   Saturday   34300   34750   34300   34680
1395-01-08 2016-03-27     Sunday   34660   34670   34660   34660
1395-01-10 2016-03-29    Tuesday   34960   35140   34960   35130
1395-01-11 2016-03-30  Wednesday   34890   34890   34820   34820
1395-01-12 2016-03-31   Thursday   34850   34850   34850   34850
...               ...        ...     ...     ...     ...     ...
1400-11-04 2022-01-24     Monday  279300  281400  278850  279790
1400-11-05 2022-01-25    Tuesday  278170  278800  277350  277760
1400-11-06 2022-01-26  Wednesday  279020  281300  278360  278610
1400-11-07 2022-01-27   Thursday  277240  278100  276860  277710
1400-11-09 2022-01-29   Saturday  282080  282800  279050  279270

[1690 rows x 6 columns]